src/com/google/javascript/jscomp/parsing/parser/util/UnicodeMatch.java

```diff
@@ -0,0 +1,174 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.javascript.jscomp.parsing.parser.util;
+
+import java.util.regex.Pattern;
+
+/**
+ * Exposes non-ASCII unicode detection for Javascript identifiers in compliance
+ * with ECMAScript spec. Retains old optimizations for commonly-used characters,
+ * namely ASCII and Latin/Greek charsets. Uses regex for remaining characters
+ * (which would throw errors otherwise).
+ * 
+ * Unicode category patterns with /u flag compiled to vanilla regex with
+ * regexpu: https://github.com/mathiasbynens/regexpu
+ * 
+ * IdentifierStart and IdentifierPart regex patterns based on ECMA's
+ * documentation:
+ * https://github.com/tc39/proposal-regexp-unicode-property-escapes#other-examples
+ * 
+ * const regexIdentifierStart = /[$_\p{ID_Start}]/u;
+ * const regexIdentifierPart = /[$_\u200C\u200D\p{ID_Continue}]/u;
+ */
+public final class UnicodeMatch {
+  private UnicodeMatch() {}
+
+  /**
+   * FAST PATHS
+   * 
+   * These fast paths are broken into distinct parts because of the logical
+   * differences in `IdentifierStart` and `IdentifierPart`, where the former
+   * does not accept a digit.
+   *
+   * A fast path is created for `UnicodeLetter` and `UnicodeDigit`, and then one
+   * each for `IdentifierStart` and `IdentifierPart`.
+   */
+
+  /**
+   * Fast path to check if a given character is a non-ASCII Unicode letter.
+   */
+  @SuppressWarnings("ShortCircuitBoolean")
+  private static final boolean fastPathUnicodeLetter(int ch) {
+    /**
+     * Intentionally avoiding short circuiting behavior of "||" and "&&". This
+     * minimizes branches in this code which minimizes branch prediction misses.
+     */
+    return (
+      // Latin letters
+      (0x00C0 <= ch & ch <= 0x00D6) |
+      // Latin letters
+      (0x00D8 <= ch & ch <= 0x00F6) |
+      // Latin letters
+      (0x00F8 <= ch & ch <= 0x00FF) |
+      // Latin Barred O
+      (ch == 0x0275) |
+      // Greek uppercase letters
+      (0x0391 <= ch & ch <= 0x03A1) |
+      // Remaining Greek uppercase letters
+      (0x03A3 <= ch & ch <= 0x03A9) |
+      // Greek lowercase letters
+      (0x03B1 <= ch & ch <= 0x03C9)
+    );
+  }
+
+  /**
+   * Fast path to check if a given character is a Unicode letter, "$" or "_".
+   */
+  @SuppressWarnings("ShortCircuitBoolean")
+  private static final boolean fastPathIdentifierStart(int ch) {
+    /**
+     * ASCII optimization. If ASCII character does not match /[a-z]/i, it is not
+     * a valid `IdentifierStart`.
+     * 
+     * Intentionally avoiding short circuiting behavior of "||" and "&&". This
+     * minimizes branches in this code which minimizes branch prediction misses.
+     */
+    if (ch <= 127) {
+      return (
+        (ch >= 'A' & ch <= 'Z') |
+        (ch >= 'a' & ch <= 'z') |
+        (ch == '$' | ch == '_')
+      );
+    }
```

Comment:
ASCII fast path will work in every call to `fastPathIdentifierStart` and `fastPathIdentifierPart` (because the former is evaluated in the latter).